src/proxy/cpsubsystem/SessionlessSemaphoreProxy.ts

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @ignore *//** */
+
+import * as Long from 'long';
+import {HazelcastClient} from '../../HazelcastClient';
+import {BaseCPProxy} from './BaseCPProxy';
+import {ISemaphore} from '../ISemaphore';
+import {CPSubsystemImpl} from '../../CPSubsystem';
+import {CPProxyManager} from './CPProxyManager';
+import {CPSessionManager, NO_SESSION_ID} from './CPSessionManager';
+import {RaftGroupId} from './RaftGroupId';
+import {assertNonNegativeNumber} from '../../util/Util';
+import {UuidUtil} from '../../util/UuidUtil';
+import {SemaphoreInitCodec} from '../../codec/SemaphoreInitCodec';
+import {SemaphoreAcquireCodec} from '../../codec/SemaphoreAcquireCodec';
+import {SemaphoreAvailablePermitsCodec} from '../../codec/SemaphoreAvailablePermitsCodec';
+import {SemaphoreDrainCodec} from '../../codec/SemaphoreDrainCodec';
+import {SemaphoreChangeCodec} from '../../codec/SemaphoreChangeCodec';
+import {SemaphoreReleaseCodec} from '../../codec/SemaphoreReleaseCodec';
+import {
+    IllegalStateError,
+    WaitKeyCancelledError
+} from '../../core';
+
+/** @internal */
+export class SessionlessSemaphoreProxy extends BaseCPProxy implements ISemaphore {
+
+    private readonly sessionManager: CPSessionManager;
+
+    constructor(client: HazelcastClient,
+                groupId: RaftGroupId,
+                proxyName: string,
+                objectName: string) {
+        super(client, CPProxyManager.SEMAPHORE_SERVICE, groupId, proxyName, objectName);
+        this.sessionManager = (client.getCPSubsystem() as CPSubsystemImpl).getCPSessionManager();
+    }
+
+    init(permits: number): Promise<boolean> {
+        assertNonNegativeNumber(permits);
+        return this.encodeInvokeOnRandomTarget(SemaphoreInitCodec, this.groupId, this.objectName, permits)
+            .then((clientMessage) => {
+                const response = SemaphoreInitCodec.decodeResponse(clientMessage);
+                return response.response;
+            });
+    }
+
+    acquire(permits = 1): Promise<void> {
+        assertNonNegativeNumber(permits);
+
+        return this.doTryAcquire(permits, -1).then();
+    }
+
+    tryAcquire(permits: number, timeout = 0): Promise<boolean> {
```

Comment:
Same comment as session aware version about defaulting permits to 1

Reply:
Good suggestion. Addressed in 2f9666008f45a3c85806531fe46524249d4a6a6c